### PR TITLE
Add outline log GRPC proto, service implementation and server skeleton

### PR DESCRIPTION
**Pull request by Martin2112**
It's still a bit WIP but I didn't think there was much point implementing operations until we decide how the pieces are going to be hooked up.

I'm starting to think that clients shouldn't use logids directly. We should give them an opaque token so they can't easily trash a different tree. Haven't implemented this yet though.

### CODE REVIEW

File: server/TrillianLogServer.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
+package server
+
+import (
+	"errors"
+
+	"github.com/google/trillian"
+	"github.com/google/trillian/storage"
+	"golang.org/x/net/context"
+	"sync"
```
Comment (AlCutter) [inline server/TrillianLogServer.go]: this up to the previous block

File: server/TrillianLogServer.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	ErrNotImplemented = errors.New("Not yet implemented")
+)
+
+// LogStorageProviderFunc decouples the server from storage implementations
+type LogStorageProviderFunc func(int64) *storage.LogStorage
+
+// TrillianLogServer implements the RPC API defined in the proto
+type TrillianLogServer struct {
+	storageProvider LogStorageProviderFunc
+	// Must hold this lock before access to the storage map
```
Comment (AlCutter) [inline server/TrillianLogServer.go]: /access to/accessing/?

File: server/TrillianLogServer.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	storageMapGuard sync.Mutex
+	// Map from storage ID to storage impl for that log
```
Comment (AlCutter) [inline server/TrillianLogServer.go]: is `storage ID` TreeID elsewhere?

File: server/TrillianLogServer.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	storageMap map[int64]*storage.LogStorage
+}
+
+// NewTrillianLogServer creates a new RPC server backed by LogStorage
```
Comment (AlCutter) [inline server/TrillianLogServer.go]: comment out of date wrt to LogStorage?

File: server/TrillianLogServer.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+)
+
+var (
+	// TODO: Delete when implementation done
+	// ErrNotImplemented is returned when an operation is not supported yet
+	ErrNotImplemented = errors.New("Not yet implemented")
+)
+
+// LogStorageProviderFunc decouples the server from storage implementations
+type LogStorageProviderFunc func(int64) *storage.LogStorage
```
Comment (AlCutter) [inline server/TrillianLogServer.go]: Could the impl of this func reasonably fail and need to return an error (unknown ID, DB unavailable, etc.)?

File: storage/tools/common.go [hunk 6]
```diff
@@ -6,18 +6,30 @@ import (
 	"github.com/google/trillian"
 	"github.com/google/trillian/storage"
 	"github.com/google/trillian/storage/mysql"
+	"github.com/google/trillian/server"
```
Comment (AlCutter) [inline storage/tools/common.go]: alpha ordering

File: storage/tools/common.go [hunk 7]
```diff
@@ -6,18 +6,30 @@ import (
...
 )
 
 var logIdFlag = flag.String("logid", "logId", "The log id to use")
 var treeIdFlag = flag.Int64("treeid", 3, "The tree id to use")
 var storageTypeFlag = flag.String("storage_type", "mysql", "Which type of storage to use")
 var mysqlUriFlag = flag.String("mysql_uri", "test:zaphod@tcp(127.0.0.1:3306)/test",
 	"uri to use with mysql storage")
+var logPortFlag = flag.Int("port", 8090, "Port to serve log requests on")
```
Comment (AlCutter) [inline storage/tools/common.go]: rename to `serverPort` or similar since this is a shared bunch of code?

File: storage/tools/common.go [hunk 8]
```diff
@@ -6,18 +6,30 @@ import (
...
 
 func GetLogIdFromFlagsOrDie() trillian.LogID {
 	return trillian.LogID{[]byte(*logIdFlag), *treeIdFlag}
 }
 
+// TODO: This needs to be tidied up
```
Comment (AlCutter) [inline storage/tools/common.go]: Missing public function docs in here for most things

File: storage/tools/common.go [hunk 9]
```diff
@@ -6,18 +6,30 @@ import (
...
+func GetLogStorageProviderFromFlags() server.LogStorageProviderFunc {
+	return func(x int64) *storage.LogStorage {
+		// TODO: We need to sort out exactly what a log id is. We really only need a tree id
+		// but having the log id as well could be a handy cross check as they can be guessed
+		storage := GetStorageFromFlagsOrDie(trillian.LogID{[]byte("This needs fixing"), x})
```
Comment (AlCutter) [inline storage/tools/common.go]: could just return it directly?

File: storage/tools/log_client/main.go [hunk 10]
```diff
@@ -0,0 +1,42 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"time"
+
+	_ "github.com/go-sql-driver/mysql"
```
Comment (AlCutter) [inline storage/tools/log_client/main.go]: Do you need this for the client?

File: server/TrillianLogServer.go (hunk 1 above)
Comment (Martin2112) [inline server/TrillianLogServer.go]: Done.

File: server/TrillianLogServer.go (hunk 2 above)
Comment (Martin2112) [inline server/TrillianLogServer.go]: Done but the other was also proper English like what she is written.

File: server/TrillianLogServer.go (hunk 3 above)
Comment (Martin2112) [inline server/TrillianLogServer.go]: Yes I meant treeID but developed a storage fixation.

File: server/TrillianLogServer.go (hunk 4 above)
Comment (Martin2112) [inline server/TrillianLogServer.go]: Done.

File: server/TrillianLogServer.go (hunk 5 above)
Comment (Martin2112) [inline server/TrillianLogServer.go]: Done.

File: storage/tools/common.go (hunk 6 above)
Comment (Martin2112) [inline storage/tools/common.go]: Done.

File: storage/tools/common.go (hunk 7 above)
Comment (Martin2112) [inline storage/tools/common.go]: Done. But I think this code won't be around too long.

File: storage/tools/common.go (hunk 8 above)
Comment (Martin2112) [inline storage/tools/common.go]: Done.

File: storage/tools/common.go (hunk 9 above)
Comment (Martin2112) [inline storage/tools/common.go]: Not sure if that works, was getting "x is pointer to interface, not interface" errors.

File: storage/tools/log_client/main.go (hunk 10 above)
Comment (Martin2112) [inline storage/tools/log_client/main.go]: No. Removed.

File: storage/tools/common.go [hunk 11]
```diff
@@ -39,12 +43,25 @@ func GetStorageFromFlagsOrDie(treeId trillian.LogID) storage.LogStorage {
...
 
-	panic(fmt.Sprintf("Unknown storage type: %s", *storageTypeFlag))
+	return nil, fmt.Errorf("Unknown storage type: %s", *storageTypeFlag)
 }
 
-func GetLogServingPort() int {
-	return *logPortFlag
-}
\ No newline at end of file
+// GetStorageFromFlags returns a configured storage instance, errors are fatal and if func init() {
```
Comment (AlCutter) [inline storage/tools/common.go]: docs vs. API name mismatch :)

Comment (AlCutter) [inline storage/tools/common.go]: "if func init() { returns" ?

Comment (Martin2112) [inline storage/tools/common.go]: Managed to trigger IDE template expansion without realizing it. Fixed.

File: storage/tools/common.go [hunk 12]
```diff
@@ -39,12 +43,25 @@ func GetStorageFromFlagsOrDie(treeId trillian.LogID) storage.LogStorage {
...
+	storage, err := GetStorageFromFlags(treeId)
+
+	if err != nil {
+		panic(err)
+	}
+
+	return storage
+}
+
+// Gets the port number to be used when serving log data
```
Comment (AlCutter) [inline storage/tools/common.go]: nit: I _think_ golint will complain about this, the Hitch Hikers^W^W style guide has this to say on the subject:

> Doc comments work best as complete sentences, which allow a wide variety of automated presentations. The first sentence should be a one-sentence summary that starts with the name being declared.

Comment (Martin2112) [inline storage/tools/common.go]: Done. Funny how Go ignores the existing conventions of every other language and then insists you follow theirs exactly. Grumble, mutter etc.

Comment (pphaneuf) [inline storage/tools/common.go]: Same convention in Python. :wink:

Comment (Martin2112) [inline storage/tools/common.go]: I rest my case ...

**Pull request merged**